src/libs/logging/src/file_rotate.rs

```diff
@@ -0,0 +1,247 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+use std::fs::{self, File, OpenOptions};
+use std::io::{self, LineWriter, Result, Write};
+use std::path::{Path, PathBuf};
+
+// Default rotate size for logger files.
+const fn default_rotate_size() -> u64 {
+    10485760
+}
+
+// Default number of log files to keep.
+const fn default_rotate_keep() -> usize {
+    3
+}
+
+/// Writer with file rotation for log files.
+///
+/// This is a modified version of `FileAppender` from
+/// https://github.com/sile/sloggers/blob/153c00a59f7218c1d96f522fb7a95c80bb0d530c/src/file.rs#L190
+#[derive(Debug)]
+pub struct FileRotator {
+    path: PathBuf,
+    file: Option<LineWriter<File>>,
+    truncate: bool,
+    written_size: u64,
+    rotate_size: u64,
+    rotate_keep: usize,
+}
+
+impl FileRotator {
+    /// Create a new instance of [`FileRotator`] to write log file at `path.
+    ///
+    /// It returns `std::io::Error` if the path is not a normal file or the parent directory does
+    /// not exist.
+    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let p = Path::new(path.as_ref());
+        if let Ok(md) = p.metadata() {
+            if !md.is_file() {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("path '{}' is not a file", p.to_string_lossy()),
+                ));
+            }
+        }
+        if let Some(parent) = p.parent() {
+            if p.has_root() || !parent.as_os_str().is_empty() {
+                let md = parent.metadata()?;
+                if !md.is_dir() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("parent of '{}' is not a directory", p.to_string_lossy()),
+                    ));
+                }
+            }
+        }
+
+        Ok(FileRotator {
+            path: path.as_ref().to_path_buf(),
+            file: None,
+            truncate: false,
+            written_size: 0,
+            rotate_size: default_rotate_size(),
+            rotate_keep: default_rotate_keep(),
+        })
+    }
+
+    /// Use "truncate" or "append" mode when opening the log file.
+    pub fn truncate_mode(&mut self, truncate: bool) -> &mut Self {
```

Comment:
When would you want `truncate_mode(true)`?  I'm not seeing any obvious use for it.

Reply:
The major usage cases should use "append" mode instead of "truncate" mode. But it may be used to get a clean log file for some trivial cases, such as debug.

Reply:
Is it worth introducing a whole mode for that, though?  Couldn't the debug style just explicitly remove the file, then open it up in append mode?

Reply:
I have no strong preference here, it's inherited from slogger.